cmd/crd-linter/linters/preserve_unknown_fields.go

```diff
@@ -0,0 +1,52 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package linters
+
+import (
+	"fmt"
+
+	v1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+)
+
+// NoPreserveUnknownFields checks if a CRD contains a 'preserveUnknownFields' marker
+// either at the top level, or on any specific sub-field of the resource.
+type NoPreserveUnknownFields struct{}
+
+var _ Linter = NoPreserveUnknownFields{}
+
+func (p NoPreserveUnknownFields) Name() string {
+	return "NoPreserveUnknownFields"
+}
+
+func (p NoPreserveUnknownFields) Description() string {
+	return "Fields should avoid using 'preserveUnknownFields' as it means any data can be persisted into the Kubernetes apiserver " +
+		"without any guards on the size and type of data. Setting this to true is no longer permitted in the 'v1' API version of " +
+		"CustomResourceDefinitions, and as such, it should not be used: https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#field-pruning"
+}
+
+func (p NoPreserveUnknownFields) Execute(crd *v1.CustomResourceDefinition) WarningList {
+	var errs []Warning
+	if crd.Spec.PreserveUnknownFields {
+		errs = append(errs, Warning{Message: "spec.preserveUnknownFields is set to 'true'"})
+	}
+	return append(errs, recurseAllSchemas(crd.Spec.Versions, func(props v1.JSONSchemaProps, path string) []Warning {
+		if props.XPreserveUnknownFields != nil && *props.XPreserveUnknownFields == true {
+			return newWarningList(fmt.Sprintf("%s.x-kubernetes-preserve-unknown-fields is set to 'true'", path))
+		}
+		return nil
+	})...)
+}
```

Comment:
The description mentions that v1 forbids it, but that seems to be true only to spec.preserveUnknownFields and not to x-kubernetes-preserve-unknown-fields on specific properties. Am I missing something?